### PR TITLE
CollectiveBroadcast Implementation. 

### PR DESCRIPTION
**Pull request by chaserileyroberts**
This PR implements `CollectiveBroadcast` for the Nvidia GPU backend. 

Most of this is just plumbing. The big feature in this PR is that async for collective broadcast is supported and enabled by default.

### CODE REVIEW

File: xla/translate/mhlo_to_hlo/mlir_hlo_to_hlo.cc [hunk 1]
```diff
@@ -834,10 +834,19 @@ bool SimplyReturnedOp(mlir::Operation* op) {
 namespace mlir {
 namespace mhlo {
 namespace {
+LogicalResult ExportXlaOp(CollectiveBroadcastOp op, OpLoweringContext ctx) {
```
Comment (GleasonK) [inline xla/translate/mhlo_to_hlo/mlir_hlo_to_hlo.cc:837]: Can you add a filecheck test to the following files, just to ensure round-tripping doesn't break:

- https://github.com/openxla/xla/blob/d239bd4230482fc02fda58bb8ac3bde3a3339c0c/xla/translate/mhlo_to_hlo/tests/export.mlir#L589
- https://github.com/openxla/xla/blob/4d7e7b96776aff48695c9e4ca64f666a5e9d5fb5/xla/translate/hlo_to_mhlo/tests/import.hlotxt#L266

**Comment by kamaljeeti** [comment]
Hi @chaserileyroberts, Any update on this PR? Please. Thank you!

**Comment by chaserileyroberts** [comment]
Sorry, it's performance review time at Nvidia so I got held up. I'll be updating this today.

File: xla/service/gpu/runtime/collectives.cc
Comment (zhenying-liu) [inline xla/service/gpu/runtime/collectives.cc:1]: Hi Chase, as far as I know, runtime will eventually go away and runtime3 will be the default and only XLA:GPU runtime (please see runtime3/README.md). Please consider to move the code from runtime to runtime3. Fortunately there are only 2 out of 76 files are related here.

Comment (chaserileyroberts) [reply xla/service/gpu/runtime/collectives.cc:1]: I have moved the implementation to runtime3

**Comment by chaserileyroberts** [comment]
Ready to continue review.

File: xla/service/hlo_dataflow_analysis.h [hunk 2]
```diff
@@ -305,6 +305,10 @@ class HloDataflowAnalysis {
   bool UpdateAllGatherStartValueSet(HloInstruction* all_gather_start);
   bool UpdateAllGatherDoneValueSet(HloInstruction* all_gather_done);
   bool UpdateAllReduceDoneValueSet(HloInstruction* all_reduce_done);
+    bool UpdateCollectiveBroadcastStartValueSet(
```
Comment (jurahul) [inline xla/service/hlo_dataflow_analysis.h:308]: Looks like this is not called from anywhere? Bugs in dataflow analysis can later lead to incorrect buffer assignment.

File: xla/hlo/ir/hlo_opcode.h [hunk 3]
```diff
@@ -75,6 +75,9 @@ namespace xla {
   V(kCholesky, "cholesky", 1)                                                  \
   V(kClamp, "clamp", 3)                                                        \
   V(kClz, "count-leading-zeros", 1)                                            \
+  V(kCollectiveBroadcast, "collective-broadcast", kHloOpcodeIsVariadic)            \
+  V(kCollectiveBroadcastStart, "collective-broadcast-start", kHloOpcodeIsVariadic) \
```
Comment (jurahul) [inline xla/hlo/ir/hlo_opcode.h:79]: I think going forward, we want to discourage addition of specific start/done ops and use the generic async-start/done ops to support this. We do that for async reduce-scatter and all-to-all for example. Please see https://github.com/openxla/xla/blob/main/docs/async_ops.md

Comment (chaserileyroberts) [reply xla/hlo/ir/hlo_opcode.h:79]: Aw ok nice, yeah that split was a pain to deal with. I'll update the PR to only use async-start/done.

File: xla/service/hlo_dataflow_analysis.h (hunk 2 above)
Comment (chaserileyroberts) [reply xla/service/hlo_dataflow_analysis.h:308]: This is the likely culprit to the e2e issue I've been seeing

**Comment by chaserileyroberts** [comment]
I've confirmed that this works with JAX. Integration here: https://github.com/google/jax/pull/19602

File: xla/hlo/ir/hlo_opcode.h (hunk 3 above)
Comment (chaserileyroberts) [reply xla/hlo/ir/hlo_opcode.h]: I've removed the kCollectiveBroadcast[Start,Done]

File: xla/service/hlo_dataflow_analysis.h (hunk 2 above)
Comment (chaserileyroberts) [reply xla/service/hlo_dataflow_analysis.h]: This was the issue! Current state of the PR has the e2e tests working.

File: xla/translate/mhlo_to_hlo/tests/export.mlir [hunk 4]
```diff
@@ -585,6 +585,20 @@ func.func @callee(%arg0: tensor<4xi32>, %arg1: tensor<4xi32>) -> (tensor<4xi32>,
...
+func.func @main(%arg0: tensor<128x32xf32>) -> tensor<128x32xf32> {
+  %0 = "mhlo.collective_broadcast"(%arg0) {
+    replica_groups = dense<[[0, 1], [2, 3]]> : tensor<2x2xi64>,
+    channel_handle = #mhlo.channel_handle<handle = 1, type = 0>
+  } : (tensor<128x32xf32>) -> tensor<128x32xf32>
+  func.return %0 : tensor<128x32xf32>
+}
+// CHECK:  ENTRY
+// CHECK:  [[ARG:%.*]] = f32[128,32] parameter(0)
+// CHECK:  ROOT [[RESULT:%.*]] = f32[128,32] collective-broadcast(f32[128,32] [[ARG]]), channel_id=1, replica_groups={{\{\{}}0,1},{2,3}}
```
Comment (GleasonK) [inline xla/translate/mhlo_to_hlo/tests/export.mlir:598]: Something seems off about this nesting.. `{{\{\{}}0,1},{2,3}}`


Can we instead use `// CHECK-SAME{LITERAL}:  replica_groups={{0,1},{2,3}}` similar to the TP 3 below this one.

File: xla/translate/hlo_to_mhlo/tests/import.hlotxt [hunk 5]
```diff
@@ -261,6 +261,17 @@ add {
   ROOT %clamp.3 = f32[4] clamp(f32[] %Arg_0.1, f32[4] %Arg_1.2, f32[] %Arg_2.3)
 }
 
+// CHECK-LABEL:  func private @test_collective_broadcast
+// CHECK-SAME:  ([[ARG:%.*]]: tensor<128x32xf32>) -> tensor<128x32xf32>
+%test_collective_broadcast (input: f32[128,32]) -> f32[128,32] {
+  %input = f32[128,32]{1,0} parameter(0)
+  // CHECK-NEXT:  "mhlo.collective_broadcast"([[ARG]]) {
+  // CHECK-SAME:    channel_handle = #mhlo.channel_handle<handle = 1, type = 0>,
+  // CHECK-SAME:    replica_groups = dense<{{\[\[}}0, 1], [2, 3]]> : tensor<2x2xi64>
```
Comment (GleasonK) [inline xla/translate/hlo_to_mhlo/tests/import.hlotxt:270]: `CHECK-SAME{LITERAL}` as mentioned above.

**Review by GleasonK — APPROVED** [review]
Reviewed:

- `xla/client/*`
- `xla/service/shape_inference.*`
- `translate/mlir_hlo/{hlo_to_mhlo,mhlo_to_hlo}/*`

The rest is fairly compiler-specific as opposed to IR specific, so will leave it to XLA devs to review. LGTM!

File: xla/client/xla_builder.h [hunk 6]
```diff
@@ -1623,6 +1632,11 @@ class XlaBuilder {
                       const std::optional<Shape>& layout,
                       std::optional<bool> use_global_device_ids, bool async);
 
+  XlaOp CollectiveBroadcastImpl(
+      XlaOp operand,
+      absl::Span<const ReplicaGroup> replica_groups,
+      const std::optional<ChannelHandle>& channel_id, bool async);
```
[the remark this answers is not in the document]
Comment (chaserileyroberts) [reply xla/client/xla_builder.h:1638]: Good point, I'll remove it.

File: xla/translate/mhlo_to_hlo/tests/export.mlir (hunk 4 above)
Comment (chaserileyroberts) [reply xla/translate/mhlo_to_hlo/tests/export.mlir:598]: Yeah, that was ported from the collective-permute test just below where it also has strange formatting for `source-target-pairs`. I'll change it.

File: xla/translate/hlo_to_mhlo/tests/import.hlotxt (hunk 5 above)
Comment (chaserileyroberts) [reply xla/translate/hlo_to_mhlo/tests/import.hlotxt]: Done.

File: xla/translate/mhlo_to_hlo/tests/export.mlir (hunk 4 above)
Comment (chaserileyroberts) [reply xla/translate/mhlo_to_hlo/tests/export.mlir]: Done

File: xla/client/xla_builder.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (chaserileyroberts) [reply xla/client/xla_builder.h]: Done.

File: xla/hlo/ir/hlo_instruction.cc [hunk 7]
```diff
@@ -1528,6 +1539,22 @@ HloInstruction::CreateAllReduceStart(
...
+    const Shape& shape, absl::Span<HloInstruction* const> operands,
+    absl::Span<const ReplicaGroup> replica_groups,
+    const std::optional<int64_t>& channel_id) {
+  return std::make_unique<HloCollectiveBroadcastInstruction>(
+      HloOpcode::kCollectiveBroadcast, shape, operands, replica_groups,
+      channel_id);
+}
+
+
+/* ================================================================== */
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instruction.cc:1557]: delete

File: xla/hlo/ir/hlo_instruction.cc [hunk 8]
```diff
@@ -1528,6 +1539,22 @@ HloInstruction::CreateAllReduceStart(
       split_dimension);
 }
 
+
+/* ==================================================================================== */
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instruction.cc:1543]: delete

File: xla/client/xla_builder.cc [hunk 9]
```diff
@@ -3595,6 +3596,37 @@ XlaOp XlaBuilder::AllToAllTuple(
...
+    }
+    if (channel_id.has_value()) {
+      instr.set_channel_id(channel_id->handle());
+    }
+
+    return AddInstruction(std::move(instr), HloOpcode::kCollectiveBroadcast, {operand});
+  });
+}
+
+//  ========================
```
Comment (ezhulenev) [inline xla/client/xla_builder.cc:3628]: delete

File: xla/client/xla_builder.cc [hunk 10]
```diff
@@ -3595,6 +3596,37 @@ XlaOp XlaBuilder::AllToAllTuple(
...
+        ShapeInference::InferCollectiveBroadcastShape({operand_shape}));
+    *instr.mutable_shape() = shape.ToProto();
+    for (const ReplicaGroup& group : replica_groups) {
+      *instr.add_replica_groups() = group;
+    }
+    if (channel_id.has_value()) {
+      instr.set_channel_id(channel_id->handle());
+    }
+
+    return AddInstruction(std::move(instr), HloOpcode::kCollectiveBroadcast, {operand});
```
Comment (ezhulenev) [inline xla/client/xla_builder.cc:3624]: 80 chars

File: xla/hlo/ir/hlo_instruction.h [hunk 11]
```diff
@@ -947,6 +947,21 @@ class HloInstruction {
       const std::optional<int64_t>& channel_id,
       const std::optional<int64_t>& split_dimension = std::nullopt);
 
+  // Creates a communication instruction that broadcasts data cross replicas.
+  // Data is sent from to the first replica id in each group to the other ids in 
+  // the same group. If a replica id is not a in any replica group, the output on that replica is a tensor
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instruction.h:952]: 80 chars

File: xla/hlo/ir/hlo_instructions.cc [hunk 12]
```diff
@@ -1101,6 +1101,29 @@ bool HloAllToAllInstruction::IdenticalSlowPathIgnoringChannelIdValues(
...
+
+std::unique_ptr<HloInstruction>
+HloCollectiveBroadcastInstruction::CloneWithNewOperandsImpl(
+    const Shape& shape, absl::Span<HloInstruction* const> new_operands,
+    HloCloneContext* /*context*/) const {
+    return std::make_unique<HloCollectiveBroadcastInstruction>(
+        opcode(), shape, new_operands, replica_groups(), channel_id());
+}
+
+// =================================================================================
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instructions.cc:1124]: delete

File: xla/hlo/ir/hlo_instructions.h [hunk 13]
```diff
@@ -449,7 +449,7 @@ class HloCholeskyInstruction : public HloInstruction {
 
 // Class that represents instructions that synchronize and transfer data between
 // partitioned devices. Send/Recv and collective instructions (AllReduce,
-// AllToAll, CollectivePermute) belong to this instruction type. A group of
+// AllToAll, CollectivePermute, CollectiveBroadcast) belong to this instruction type. A group of
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instructions.h:452]: 80 chars

File: xla/hlo/ir/hlo_instructions.h [hunk 14]
```diff
@@ -847,6 +847,29 @@ class HloAllToAllInstruction : public HloCollectiveInstruction {
   std::optional<int64_t> split_dimension_;
 };
 
+
+class HloCollectiveBroadcastInstruction : public HloCollectiveInstruction {
+ public:
+  explicit HloCollectiveBroadcastInstruction(
+      HloOpcode opcode, const Shape& shape, absl::Span<HloInstruction* const> operands,
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_instructions.h:854]: 80 chars

File: xla/hlo/ir/hlo_opcode.h [hunk 15]
```diff
@@ -75,6 +75,7 @@ namespace xla {
   V(kCholesky, "cholesky", 1)                                                  \
   V(kClamp, "clamp", 3)                                                        \
   V(kClz, "count-leading-zeros", 1)                                            \
+  V(kCollectiveBroadcast, "collective-broadcast", kHloOpcodeIsVariadic)            \
```
Comment (ezhulenev) [inline xla/hlo/ir/hlo_opcode.h:78]: 80 chars

File: xla/hlo/utils/hlo_query.cc [hunk 16]
```diff
@@ -31,6 +31,7 @@ namespace hlo_query {
 bool IsCollectiveCommunicationOp(HloOpcode op) {
   return op == HloOpcode::kAllReduce || op == HloOpcode::kAllGather ||
          op == HloOpcode::kAllToAll || op == HloOpcode::kCollectivePermute ||
+         op == HloOpcode::kCollectiveBroadcast ||
```
Comment (ezhulenev) [inline xla/hlo/utils/hlo_query.cc:34]: I'm pretty clang-format will complain

File: xla/service/cpu/ir_emitter.cc [hunk 17]
```diff
@@ -1448,6 +1448,46 @@ Status IrEmitter::HandleAllGather(HloInstruction* instruction) {
   return OkStatus();
 }
 
+Status IrEmitter::HandleCollectiveBroadcast(HloInstruction* crs) {
```
Comment (ezhulenev) [inline xla/service/cpu/ir_emitter.cc:1451]: HloInstruction* inst (or hlo)

File: xla/service/gpu/buffer_sharing.cc [hunk 18]
```diff
@@ -203,6 +203,7 @@ std::optional<bool> CanShareBufferHint(const HloInstruction* user,
                                        const ShapeIndex& user_index) {
   switch (user->opcode()) {
     case HloOpcode::kAllReduce:
+    case HloOpcode::kCollectiveBroadcast:
       // NCCL all-reduce can be performed in-place.
```
Comment (ezhulenev) [inline xla/service/gpu/buffer_sharing.cc:207]: update comment

File: xla/service/gpu/gpu_convert_async_collectives_to_sync_test.cc [hunk 19]
```diff
@@ -187,7 +207,7 @@ TEST_F(GpuConvertAsyncCollectivesToSyncTest, SimpleReduceScatter) {
 TEST_F(GpuConvertAsyncCollectivesToSyncTest, SimpleAllToAll) {
   const absl::string_view hlo_string = R"(
   HloModule test, is_scheduled=true
-
+
```
Comment (ezhulenev) [inline xla/service/gpu/gpu_convert_async_collectives_to_sync_test.cc:210]: empty line

File: xla/service/gpu/ir_emitter_unnested.cc [hunk 20]
```diff
@@ -4482,6 +4483,31 @@ absl::Status IrEmitterUnnested::EmitOp(
...
 
+  if (mlir::isa<mlir::lmhlo_gpu::CollectiveBroadcastStartOp>(op)) {
+    if (ir_emitter_context_->emit_ir_from_hlo()) 
+    {
+      auto* async_start = hlo_for_lmhlo.at(op);
+      auto* collective_broadcast = Cast<HloCollectiveBroadcastInstruction>(
+          async_start->async_wrapped_instruction());
+      return EmitNcclThunk<NcclCollectiveBroadcastStartThunk,
+                           HloCollectiveBroadcastInstruction>(
+          Thunk::kNcclCollectiveBroadcastStart, async_start, collective_broadcast, std::nullopt);
```
Comment (ezhulenev) [inline xla/service/gpu/ir_emitter_unnested.cc:4494]: 80 chars

File: xla/service/gpu/nccl_api.h [hunk 21]
```diff
@@ -179,6 +179,14 @@ class NcclApi {
                                  ReductionKind reduction_kind,
                                  NcclCommHandle comm, se::Stream* stream) = 0;
 
+  // Copy data in `send_buff` from the root GPU to the `recv_buff` on 
+  // all GPUs.
+  //
+  // https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/colls.html#ncclbroadcast
+  static absl::Status Broadcast(se::DeviceMemoryBase send_buffer,
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_api.h:186]: 80 chars

Comment (ezhulenev) [inline xla/service/gpu/nccl_api.h:186]: this also has to be added to nccl_api_stub.cc

File: xla/service/gpu/nccl_collective_broadcast_thunk.cc [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+#include "absl/container/flat_hash_set.h"
+#include "xla/mlir_hlo/lhlo_gpu/IR/lhlo_gpu_ops.h"
+#include "xla/service/collective_ops_utils.h"
+#include "xla/service/gpu/ir_emission_utils.h"
+#include "xla/translate/mhlo_to_hlo/attribute_exporter.h"
+#include "xla/xla_data.pb.h"
+#include "xla/service/gpu/nccl_api.h"
+
+
+#if XLA_ENABLE_XCCL
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_collective_broadcast_thunk.cc:33]: this should not be needed, I'm working on removing this ifdef from XLA

File: xla/service/gpu/nccl_collective_broadcast_thunk.cc [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+#include "xla/stream_executor/gpu/gpu_stream.h"
+#endif
+
+namespace xla {
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_collective_broadcast_thunk.cc:37]: namespace xla::gpu

File: xla/service/gpu/nccl_collective_broadcast_thunk.h [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+#include "xla/service/gpu/nccl_collective_thunk.h"
+#include "xla/service/gpu/nccl_api.h"
+
+namespace xla {
+namespace gpu {
+// Thunk that performs a NCCL-based collective broadcast.
+class NcclCollectiveBroadcastStartThunk : public NcclCollectiveThunk {
+ public:
+
+  static Status CheckImplementable(mlir::lmhlo_gpu::CollectiveBroadcastStartOp op,
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_collective_broadcast_thunk.h:31]: constructors from MLIR can be removed, I plan to start deleting LMHLO soon

File: xla/service/gpu/nccl_collective_broadcast_thunk.h [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+#ifndef XLA_SERVICE_GPU_NCCL_COLLECTIVE_BROADCAST_THUNK_H_
+#define XLA_SERVICE_GPU_NCCL_COLLECTIVE_BROADCAST_THUNK_H_
+
+#include <cstdint>
+
+#include "xla/service/collective_ops_utils.h"
+#include "xla/service/gpu/nccl_collective_thunk.h"
+#include "xla/service/gpu/nccl_api.h"
+
+namespace xla {
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_collective_broadcast_thunk.h:25]: namespace xla::gpu

File: xla/service/gpu/nccl_collective_broadcast_thunk.cc [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+  TF_ASSIGN_OR_RETURN(
+      std::vector<DeviceBufferPair> device_buffers,
+      ConvertToDeviceBuffers(params, buffers_,
+                             config_.operand_element_type));
+  return ::xla::gpu::RunCollectiveBroadcast(device_buffers, stream, comm, nccl_api());
+}
+
+Status RunCollectiveBroadcast(std::vector<DeviceBufferPair>& buffers, se::Stream& stream,
+                            NcclApi::NcclCommHandle comm, NcclApi* nccl_api) {
+#if XLA_ENABLE_XCCL
```
Comment (ezhulenev) [inline xla/service/gpu/nccl_collective_broadcast_thunk.cc:106]: this is also not needed, everything in XLA should work these days without a need to hide behind XLA_ENABLE_XCCL

File: xla/service/gpu/runtime/command_buffer_cmd.cc [hunk 27]
```diff
@@ -1317,4 +1318,70 @@ CommandBufferCmd::BufferUsageVector AllGatherCmd::buffers() {
...
+
+absl::Status CollectiveBroadcastCmd::Record(const Thunk::ExecuteParams& params,
+                                  StateManager& state,
+                                  se::CommandBuffer* command_buffer) {
+  TF_ASSIGN_OR_RETURN(
+      std::vector<DeviceBufferPair> device_buffers,
+      ConvertToDeviceBuffers(params.buffer_allocations, buffers_,
+                             config().operand_element_type));
+
+  VLOG(5) << "CollectiveBroadcasteCmd:";
```
Comment (ezhulenev) [inline xla/service/gpu/runtime/command_buffer_cmd.cc:1339]: typo

File: xla/service/gpu/nccl_collective_broadcast_thunk.cc (hunk 26 above)
Comment (chaserileyroberts) [reply xla/service/gpu/nccl_collective_broadcast_thunk.cc:106]: Good to know, I will remove it

File: xla/service/gpu/buffer_sharing.cc (hunk 18 above)
Comment (chaserileyroberts) [reply xla/service/gpu/buffer_sharing.cc:207]: Done

File: xla/service/gpu/gpu_convert_async_collectives_to_sync_test.cc (hunk 19 above)
Comment (chaserileyroberts) [reply xla/service/gpu/gpu_convert_async_collectives_to_sync_test.cc:210]: Done

File: xla/service/gpu/nccl_api.h (hunk 21 above)
Comment (chaserileyroberts) [reply xla/service/gpu/nccl_api.h:186]: added

File: xla/service/gpu/nccl_collective_broadcast_thunk.cc (hunk 22 above)
Comment (chaserileyroberts) [reply xla/service/gpu/nccl_collective_broadcast_thunk.cc:33]: removed.

File: xla/hlo/ir/hlo_instructions.cc (hunk 12 above)
Comment (chaserileyroberts) [reply xla/hlo/ir/hlo_instructions.cc:1124]: removed.

File: xla/hlo/ir/hlo_instruction.cc (hunk 8 above)
Comment (chaserileyroberts) [reply xla/hlo/ir/hlo_instruction.cc:1543]: removed.

File: xla/client/xla_builder.cc (hunk 9 above)
Comment (chaserileyroberts) [reply xla/client/xla_builder.cc:3628]: removed.

File: xla/hlo/ir/hlo_instruction.cc (hunk 7 above)
Comment (chaserileyroberts) [reply xla/hlo/ir/hlo_instruction.cc:1557]: removed

File: xla/mlir/backends/gpu/transforms/lmhlo_to_gpu_runtime.cc
Comment (ghpvnist) [inline xla/mlir/backends/gpu/transforms/lmhlo_to_gpu_runtime.cc:1]: This is an empty file. Is this meant to be included?

Comment (chaserileyroberts) [reply xla/mlir/backends/gpu/transforms/lmhlo_to_gpu_runtime.cc:1]: No, I think this was a holdover from some rebase. Removing.

File: xla/hlo/utils/hlo_query.cc (hunk 16 above)
Comment (chaserileyroberts) [reply xla/hlo/utils/hlo_query.cc:34]: What do you mean by this?

File: xla/service/gpu/nccl_collective_broadcast_thunk.h (hunk 24 above)
Comment (chaserileyroberts) [reply xla/service/gpu/nccl_collective_broadcast_thunk.h]: I've removed all methods that use `lmhlo_gpu::CollectiveBroadcastStartOp`.

**Comment by chaserileyroberts** [comment]
@ezhulenev I believe I've addressed all of your comments. This is ready for the next round of reviews.

**Review by ezhulenev — APPROVED** [review]

File: xla/service/hlo_verifier.cc [hunk 28]
```diff
@@ -762,6 +762,15 @@ Status CheckDuplicatedSourceOrTarget(HloInstruction* hlo,
 
 }  // namespace
 
+Status ShapeVerifier::HandleCollectiveBroadcast(HloInstruction* hlo) {
```
[the remark this answers is not in the document]
Comment (chaserileyroberts) [reply xla/service/hlo_verifier.cc:765]: I think I can include that now actually. The implementation of the thunk should already support it, so I'd just need to flip this flag and add a test. Let me do that.

**Review by jurahul — APPROVED** [review]

**Review by jurahul — APPROVED** [review]

**Review by jurahul — APPROVED** [review]

**Review by ddunl — APPROVED** [review]

File: xla/service/gpu/runtime/command_buffer_cmd.h [hunk 29]
```diff
@@ -784,6 +784,24 @@ class AllGatherCmd : public CollectiveCmd {
...
+// CollectiveBroadcastCmd
+//===----------------------------------------------------------------------===//
+
+class CollectiveBroadcastCmd : public CollectiveCmd {
+ public:
+  CollectiveBroadcastCmd(NcclApi* nccl_api, NcclCollectiveConfig config,
+                         absl::Span<const NcclCollectiveThunk::Buffer> buffers);
+
+  absl::Status Record(const Thunk::ExecuteParams& params, StateManager& state,
+                      se::CommandBuffer* command_buffer) override;
```
Comment (thomasjoerg) [inline xla/service/gpu/runtime/command_buffer_cmd.h:797]: I see several build breakages, e.g. `xla/service/gpu:ir_emitter_unnested` all failing with 
`error: non-virtual member function marked 'override' hides virtual member function
se::CommandBuffer* command_buffer) override;`

Comment (chaserileyroberts) [reply xla/service/gpu/runtime/command_buffer_cmd.h:797]: Can I get my context here? I'm not sure what to fix.

**Review by ddunl — APPROVED** [review]

**Review by ddunl — APPROVED** [review]

File: xla/service/gpu/runtime/command_buffer_cmd.h (hunk 29 above)
Comment (chaserileyroberts) [reply xla/service/gpu/runtime/command_buffer_cmd.h:904]: Fixed.

**Comment by chaserileyroberts** [comment]
Whats the feedback from the internal pull?

File: xla/hlo/ir/hlo_instructions.cc [hunk 30]
```diff
@@ -1118,6 +1118,26 @@ bool HloAllToAllInstruction::IdenticalSlowPathIgnoringChannelIdValues(
          split_dimension_ == casted_other.split_dimension();
 }
 
+HloCollectiveBroadcastInstruction::HloCollectiveBroadcastInstruction(
+    HloOpcode opcode, const Shape& shape,
+    absl::Span<HloInstruction* const> operands,
+    absl::Span<const ReplicaGroup> replica_groups,
+    const std::optional<int64_t>& channel_id)
+    : HloCollectiveInstruction(opcode, shape, operands, replica_groups,
+                               constrain_layout(), channel_id) {}
```
Comment (jurahul) [inline xla/hlo/ir/hlo_instructions.cc:1127]: This is the cause of ubsan failure. constrain_layout() is uniitialized. May be this just needs to be false.

**Review by ddunl — APPROVED** [review]

**Review by ddunl — APPROVED** [review]

**Review by ddunl — APPROVED** [review]

**Comment by cheshire** [comment]
Apologies for conflicting review information, but 65 files is really a lot for one commit.

Could we decouple most of those?

E.g. add HLO first (which crashes in a verifier at runtime), then implementation, then bridge changes? also @klucke

**Comment by jurahul** [comment]
@chaserileyroberts can we abandon this PR now that it has been split and merged?

**Comment by chaserileyroberts** [comment]
Yes. 

God speed gentlemen.

**Pull request closed without merging**